File: python/multifile/validators.py
"""Input validation utilities."""

import re
from typing import Optional


def validate_user_id(user_id: str) -> bool:
    """Validate that user_id is a positive integer string.

    Matches template: py-multifile-sqli-safe (validator)
    """
    if not user_id:
        return False

    # Must be numeric only
    if not user_id.isdigit():
        return False

    # Must be positive
    if int(user_id) <= 0:
        return False

    return True


def sanitize_search_query(query: str) -> str:
    """Sanitize search query by removing special characters.

    Matches template: py-multifile-sqli-search-safe (sanitizer)
    """
    if not query:
        return ""

    # Remove SQL special characters
    # This is defense-in-depth; parameterized queries are the primary protection
    sanitized = re.sub(r"[;'\"\\]", "", query)

    # Limit length
    return sanitized[:100]


def validate_email(email: str) -> bool:
    """Validate email format."""
    if not email:
        return False

    # Basic email pattern
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return bool(re.match(pattern, email))


def validate_filename(filename: str) -> bool:
    """Validate filename doesn't contain path traversal.

    Matches template: py-multifile-pathtraversal-safe (validator)
    """
    if not filename:
        return False

    # No path separators
    if '/' in filename or '\\' in filename:
        return False

    # No parent directory references
    if '..' in filename:
        return False

    # Only alphanumeric, dash, underscore, and single dot
    pattern = r'^[a-zA-Z0-9_-]+(\.[a-zA-Z0-9]+)?$'
    return bool(re.match(pattern, filename))


def validate_command_key(key: str, allowed_keys: set) -> bool:
    """Validate command key is in allowlist.

    Matches template: py-multifile-cmdi-safe (validator)
    """
    return key in allowed_keys

File: typescript/multifile/validators.ts
/**
 * Input validation utilities.
 */

/**
 * Validate that userId is a positive integer string.
 *
 * Matches template: ts-multifile-sqli-safe (validator)
 */
export function validateUserId(userId: string): boolean {
  if (!userId) {
    return false;
  }

  // Must be numeric only
  if (!/^\d+$/.test(userId)) {
    return false;
  }

  // Must be positive
  const num = parseInt(userId, 10);
  if (num <= 0 || !Number.isFinite(num)) {
    return false;
  }

  return true;
}

/**
 * Sanitize search query by removing special characters.
 *
 * Matches template: ts-multifile-sqli-search-safe (sanitizer)
 */
export function sanitizeSearchQuery(query: string): string {
  if (!query) {
    return '';
  }

  // Remove SQL special characters
  // This is defense-in-depth; parameterized queries are the primary protection
  const sanitized = query.replace(/[;'"\\]/g, '');

  // Limit length
  return sanitized.slice(0, 100);
}

/**
 * Validate email format.
 */
export function validateEmail(email: string): boolean {
  if (!email) {
    return false;
  }

  // Basic email pattern
  const pattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  return pattern.test(email);
}

/**
 * Validate filename doesn't contain path traversal.
 *
 * Matches template: ts-multifile-pathtraversal-safe (validator)
 */
export function validateFilename(filename: string): boolean {
  if (!filename) {
    return false;
  }

  // No path separators
  if (filename.includes('/') || filename.includes('\\')) {
    return false;
  }

  // No parent directory references
  if (filename.includes('..')) {
    return false;
  }

  // Only alphanumeric, dash, underscore, and single dot
  const pattern = /^[a-zA-Z0-9_-]+(\.[a-zA-Z0-9]+)?$/;
  return pattern.test(filename);
}

/**
 * Sanitize HTML to prevent XSS.
 *
 * Matches template: ts-multifile-xss-safe (sanitizer)
 */
export function sanitizeHtml(html: string): string {
  if (!html) {
    return '';
  }

  return html
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;');
}

/**
 * Validate command key is in allowlist.
 */
export function validateCommandKey(key: string, allowedKeys: Set<string>): boolean {
  return allowedKeys.has(key);
}
